File: Service/StartService/app/src/main/java/com/example/syp/startservice/MyService.java
package com.example.syp.startservice;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import android.support.annotation.IntDef;

/**
 * Service
 * <p>
 * <p>
 * onStartCommand()方法：
 * 启动service的时候，onCreate方法只有第一次会调用，onStartCommand和onStart每次都被调用。onStartCommand会告诉系统如何重启服务，如判断是否异常终止后重新启动，在何种情况下异常终止
 * onStartComand使用时，返回的是一个(int)整形。
 * 这个整形可以有四个返回值：start_sticky、start_no_sticky、START_REDELIVER_INTENT、START_STICKY_COMPATIBILITY。
 */
public class MyService extends Service {

    public interface OnCountChangeListener{
        void onChange(int count);
    }


    private boolean running = false;
    private int count = 0;


    private OnCountChangeListener onCountChangeListener;


    public MyService() {
    }

    @Override
    public IBinder onBind(Intent intent) {

        return new MyBinder();
    }

    public class MyBinder extends Binder {
        public void SetCount(int count) {
            MyService.this.setCount(count);
        }

        public int getCount() {
            return MyService.this.getCount();
        }

        public void setOnCountChangeListener(OnCountChangeListener listener){
            MyService.this.setOnCountChangeListener(listener);
        }
    }

    @Override
    public void onCreate() {
        super.onCreate();

        running = true;

        new Thread() {
            @Override
            public void run() {
                super.run();

                while (running) {
                    try {
                        sleep(1000);

                        count++;

                        if (getOnCountChangeListener() != null) {
                            getOnCountChangeListener().onChange(count);
                        }

                        System.out.println(count);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }.start();

        System.out.println("MyService.onCreate");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        System.out.println("MyService.onStartCommand");
        int count = intent.getIntExtra("count", 0);
        if (count != 0) {
            this.count = count;
        }

        return super.onStartCommand(intent, flags, startId);
    }

    //自己写的一个get set方法
    public int getCount() {
        return count;
    }

    public void setCount(int count) {
        this.count = count;
    }


    public void setOnCountChangeListener(OnCountChangeListener onCountChangeListener) {
        this.onCountChangeListener = onCountChangeListener;
    }

    public OnCountChangeListener getOnCountChangeListener() {
        return onCountChangeListener;
    }

    @Override
    public void onDestroy() {
        System.out.println("MyService.onDestroy");
        super.onDestroy();

        running = false;
    }

}

File: Service/AIDLCallBack/anotherapp/src/main/java/com/example/anotherapp/MyService.kt
package com.example.anotherapp

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.os.RemoteCallbackList

/**
 * Created by syp on 17-5-19.
 * 2.AIDL进程间通信
 */
open class MyService : Service() {

    private var count = 0
    private var running = false
    private var countChangeListeners = RemoteCallbackList<ICountChangeListener>()


    override fun onBind(intent: Intent?): IBinder {
        return MyBinder()
    }

    inner class MyBinder : IRemoteBinder.Stub() {
        override fun registerCountChangeListener(listener: ICountChangeListener?) {
            this@MyService.registerCountChangeListener(listener)
        }

        //ICountChangeListener?  问号是防止空指针
        override fun unregisterCountChangeListener(listener: ICountChangeListener?) {
            this@MyService.unregisterCountChangeListener(listener)
        }


        override fun getCount(): Int {
            return this@MyService.count
        }

        override fun setCount(count: Int) {
            this@MyService.count = count
        }
    }

    override fun onCreate() {
        super.onCreate()

        running = true
        println("====================MyService.onCreate")

        Thread(Runnable {

            while (running) {
                count++

                countChangeListeners.beginBroadcast()

                var index = 0
                while (index < countChangeListeners.registeredCallbackCount) {
                    countChangeListeners.getBroadcastItem(index).onChange(count)
                    index++
                }

                countChangeListeners.finishBroadcast()

                println(count)

                Thread.sleep(1000)
            }
        }).start()
    }

    override fun onDestroy() {
        super.onDestroy()
        running = false

        println("====================MyService.onDestroy")
    }

    fun registerCountChangeListener(listener: ICountChangeListener?) {
        countChangeListeners.register(listener)
    }

    fun unregisterCountChangeListener(listener: ICountChangeListener?) {
        countChangeListeners.unregister(listener)
    }
}